packages/jsdom-crawler/src/internals/jsdom-crawler.ts

```diff
@@ -0,0 +1,1023 @@
+import { addTimeoutToPromise, tryCancel } from '@apify/timeout';
+import { concatStreamToBuffer, readStreamToString } from '@apify/utilities';
+import {
+    BasicCrawler,
+    BasicCrawlerOptions,
+    BASIC_CRAWLER_TIMEOUT_BUFFER_SECS,
+} from '@crawlee/basic';
+import {
+    CrawlerExtension,
+    CrawlerHandleFailedRequestInput,
+    CrawlingContext,
+    enqueueLinks,
+    EnqueueLinksOptions,
+    mergeCookies,
+    ProxyConfiguration,
+    ProxyInfo,
+    Request,
+    RequestQueue,
+    resolveBaseUrl,
+    Session,
+    storage,
+    validators,
+} from '@crawlee/core';
+import { Awaitable, Dictionary, entries, parseContentTypeFromResponse } from '@crawlee/utils';
+import contentTypeParser, { RequestLike, ResponseLike } from 'content-type';
+import { gotScraping, OptionsInit, Method, TimeoutError, Request as GotRequest } from 'got-scraping';
+import { IncomingHttpHeaders, IncomingMessage } from 'http';
+import iconv from 'iconv-lite';
+import ow from 'ow';
+import util from 'util';
+import { JSDOM, DOMWindow } from 'jsdom';
+
+/**
+ * Default mime types, which CheerioScraper supports.
```

Comment:
So... I just copy-pasted the cheerio crawler and did some changes. I was thinking of renaming `cheerio` to `jsdom` everywhere but the share is just large. I'd appreciate any comment on how to solve this.

Those are basically the same crawlers with the difference that it's just `jsdom` instead of `cheerio`. However there's more to it:

JSDOM already performs encoding detection and decodes the buffer properly (just make sure we're passing a buffer), so we could remove the code that's responsible for this. For now we just read the stream and decode it ourselves, like in the cheerio crawler.

Also, JSDOM doesn't support streaming. Here's some CloudFlare's research on HTML parsers:
https://blog.cloudflare.com/html-parsing-1/
https://blog.cloudflare.com/html-parsing-2/

I don't think lack of streaming is of that important issues. In our case, more time is spent on downloading than parsing.
Although if we used WebAssembly then the parse overhead would be almost 0ms.

Reply:
as [the issue description](https://github.com/apify/apify-ts/issues/59) says (and it is a single sentence), there should be a common predecessor for cheerio and jsdom crawlers (it could be a new package, or as long as it does not require additional dependencies, we could also add it to the `@crawlee/basic` package). it should be above basic crawler most probably, but i can as well imagine we move some methods down to the basic crawler if it makes sense. keep in mind that basic crawler is a separate unit, not an abstract class, so it needs to stay working on its own (we cant define abstract methods/props and force users to extend it)

we cant reuse cheerio interfaces here, even if they would be the same, we should use type aliases or interface extensions. reusing them would be very confusing for users as well as for us later during maintenance.

regarding webassmebly, i can imagine benefits, but also huge hickups during maintenance, especially when others will need to do changes in the code they havent written. lets optimize when we see it makes sense.